src/Components/test/testassets/BasicTestApp/JavaScriptRootComponents.razor

```diff
@@ -0,0 +1,111 @@
+﻿<h1>JavaScript root components</h1>
+
+<p>This shows that we can dynamically add root components using JavaScript.</p>
+
+<p id="message"></p>
+
+<div id="container-rendered-by-blazor"></div>
+
+<p>
+    <button id="add-root-component" onclick="addRootComponent(false)">Add new root component</button>
+    <button id="add-root-component-inside-blazor" onclick="addRootComponent(true)">Add new root component inside Blazor UI</button>
+    <button id="remove-root-component" onclick="removeRootComponent()">Remove root component</button>
+</p>
+
+<p>
+    Increment amount: <input type="number" id="increment-amount" value="1" />
+    <button id="set-increment-amount" onclick="setIncrementAmount()">Set</button>
+</p>
+
+<p>
+    <button id="set-complex-params" onclick="setComplexParameters()">Set complex parameters</button>
+    <button id="set-invalid-params" onclick="setInvalidParameters()">Set invalid parameters</button>
+    <button id="set-catchall-params" onclick="setCatchAllParameters()">Set catch-all parameters</button>
+</p>
+
+<script suppress-error="BL9992">
+    let numAddedComponents = 0;
+    const addedComponents = [];
+    const jsObject = DotNet.createJSObjectReference({
+        getValue() {
+            return `You've added ${numAddedComponents} components.`;
+        },
+
+        willDispose() {
+            setMessage('Received call to willDispose');
+        }
+    });
+
+    async function addRootComponent(intoBlazorUi) {
+        let containerElement;
+        if (intoBlazorUi) {
+            // We want to test putting it *directly* into a Blazor-rendered element, not some child of it
+            containerElement = document.getElementById('container-rendered-by-blazor');
+        } else {
+            containerElement = document.createElement('div');
+            containerElement.id = `root-container-${++numAddedComponents}`;
+            document.body.appendChild(containerElement);
+        }
+
+        const component = await Blazor.rootComponents.add(containerElement, 'my-dynamic-root-component', { incrementAmount: 1 });
```

Comment:
Can you file an issue to ensure we can do things like pass JSObjectReferences or byte[] as part of fixing the serialization?

Reply:
Added a note to https://github.com/dotnet/aspnetcore/issues/34338 because that's when I intend to address this.